### ISSUE TITLE
Potential misleading doc string for _get_train_sampler()

### ISSUE THREAD

**Issue by yuh8** [huggingface/trl#3327]
### Reproduction

Hi thanks for the amazing work on speeding up training with smart batching design. 

I think there is an error in the docstring of `_get_train_sampler()` describing how batches are splitted within devices and across devices. @qgallouedec 

Based on the specified example, for a `per_device_batch_size=3`, `num_grad_accum=2` and `num_iterations = 2`, each device should receive an `effective_batch_size =  per_device_batch_size * num_grad_accum = 6`, not what docstring has shown to be 3. The default setting of `split_batches=False` used by `accelerator.prepare()` also ensures that each device receives `effective_batch_size=6`

And then `split_tensor_dict()` method within each device/process divides this batch into two chunks of 3. So the correct docstring illustration of the process should be 

        #                                      |         GPU 0         |         GPU 1         |
        #                                      |  Accum 0  |  Accum 1  |  Accum 0  |  Accum 1  |
        #
        #                 global_step   step    <-───>  num_generations=2
        #                                       <-───────> per_device_train_batch_size=3
        #  grad_accum    ▲  ▲  0          0     [0   0   1   1   2   2]  3   3   4   4   5   5    <- Generate for the whole accumulated batch; store the completions; use the first slice to compute the loss
        #     =2         ▼  |  0          1      0   0   1   1   2   2 [ 3   3   4   4   5   5]   <- Take the stored generations and use the second slice to compute the loss
        #                   |
        #                   |  1          2     [0   0   1   1   2   2]  3   3   4   4   5   5    <- Take the stored generations and use the first slice to compute the loss
        #  num_iterations=2 ▼  1          3      0   0   1   1   2   2 [ 3   3   4   4   5   5]   <- Take the stored generations and use the second slice to compute the loss
        #
        #                      2          4     [6   6   7   7   8   8]  9   9  10  10  11  11    <- Generate for the whole accumulated batch; store the completions; use the first slice to compute the loss
        #                      2          5      6   6   7   7   8   8 [ 9   9  10  10  11  11]   <- ...
        #                                          ...

### System Info

NIL

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by LeonEricsson** [comment]
`per_device_train_batch_size` represents the samples seen by each device during a `step`, not a global step. It does not depend on the gradient accumulation steps.

**Comment by yuh8** [comment]
> `per_device_train_batch_size` represents the samples seen by each device during a `step`, not a global step. It does not depend on the gradient accumulation steps.

Thanks @LeonEricsson, you are right to point out that the `per_device_train_batch_size` is independent of accum steps. I made a mistake of using this term, rather it should be `per_device_effective_batch_size`. The `per_device_effective_batch_size` is the actual batch size each device sees during training, as evidenced by [this line ](https://github.com/huggingface/trl/blob/294f35bf3c0043d3ee6b9b5d22385e5736f6ce9e/trl/trainer/grpo_trainer.py#L684) in the grpo_trainer.py module. Each DataLoader per device has a `batch_size = self._train_batch_size * self.args.gradient_accumulation_steps` where `self._train_batch_size` references to `per_device_train_batch_size`. So effective batch size does depend on accum steps

I have changed the `per_device_train_batch_size` to `per_device_effective_batch_size` in the figure above.
